core/src/main/java/org/apache/iceberg/view/BaseViewCatalog.java

```diff
@@ -0,0 +1,190 @@
+    private final List<ViewRepresentation> viewRepresentations = Lists.newArrayList();
+    private final Map<String, String> properties = Maps.newHashMap();
+    private Schema schema;
+
+    protected BaseViewBuilder(TableIdentifier identifier) {
+      Preconditions.checkArgument(
+          isValidIdentifier(identifier), "Invalid view identifier: %s", identifier);
+      this.identifier = identifier;
+    }
+
+    @Override
+    public ViewBuilder withSchema(Schema newSchema) {
+      this.schema = newSchema;
+      viewVersionBuilder.schemaId(newSchema.schemaId());
+      return this;
+    }
+
+    @Override
+    public ViewBuilder withQuery(String dialect, String sql) {
+      viewRepresentations.add(
+          ImmutableSQLViewRepresentation.builder().dialect(dialect).sql(sql).build());
+      return this;
+    }
+
+    @Override
+    public ViewBuilder withDefaultCatalog(String defaultCatalog) {
+      viewVersionBuilder.defaultCatalog(defaultCatalog);
+      return this;
+    }
+
+    @Override
+    public ViewBuilder withDefaultNamespace(Namespace namespace) {
+      viewVersionBuilder.defaultNamespace(namespace);
+      return this;
+    }
+
+    @Override
+    public ViewBuilder withProperties(Map<String, String> newProperties) {
+      this.properties.putAll(newProperties);
+      return this;
+    }
+
+    @Override
+    public ViewBuilder withProperty(String key, String value) {
+      this.properties.put(key, value);
+      return this;
+    }
+
+    @Override
+    public View create() {
+      ViewOperations ops = newViewOps(identifier);
+      if (null != ops.current()) {
+        throw new AlreadyExistsException("View already exists: %s", identifier);
+      }
+
+      ViewVersion viewVersion =
+          viewVersionBuilder
+              .versionId(1)
+              .addAllRepresentations(viewRepresentations)
+              .timestampMillis(System.currentTimeMillis())
+              .putSummary("operation", "create")
+              .build();
+
+      ViewMetadata viewMetadata =
+          ViewMetadata.builder()
+              .setProperties(properties)
+              .setLocation(defaultWarehouseLocation(identifier))
+              .setCurrentVersion(viewVersion, schema)
+              .build();
+
+      try {
+        ops.commit(null, viewMetadata);
+      } catch (CommitFailedException ignored) {
+        throw new AlreadyExistsException("View was created concurrently: %s", identifier);
+      }
+
+      return new BaseView(ops, ViewUtil.fullViewName(name(), identifier));
+    }
+
+    @Override
+    public View replace() {
+      ViewOperations ops = newViewOps(identifier);
+      if (null == ops.current()) {
+        throw new NoSuchViewException("View does not exist: %s", identifier);
+      }
+
+      ViewMetadata metadata = ops.current();
+
+      ViewVersion viewVersion =
+          viewVersionBuilder
+              .versionId(metadata.currentVersionId() + 1)
```

Comment:
This needs to be `max(versions.map(v -> v.versionId())) + 1`. There's no guarantee that the current version has the highest version number.

Reply:
indeed, I haven't considered that here, thanks. Fixed now